src/v/transform/transform_processor.cc

```diff
@@ -0,0 +1,159 @@
+        }
+        return copied;
+    }
+}
+
+ss::future<std::optional<model::offset>> consume_batches(
+  std::unique_ptr<model::record_batch_reader::impl> reader,
+  ss::noncopyable_function<ss::future<>(model::record_batch_reader::data_t)> cb,
+  ss::abort_source* as) {
+    std::optional<model::offset> latest_offset;
+    while (!reader->is_end_of_stream() && !as->abort_requested()) {
+        // TODO(rockwood): Can this interface optionally take an abort source
+        // instead of a timeout?
+        auto slice = co_await reader->do_load_slice(model::no_timeout);
+        auto batches = extract_batches(std::move(slice));
+        if (batches.empty()) {
+            continue;
+        }
+        latest_offset = model::next_offset(batches.back().last_offset());
+        co_await cb(std::move(batches));
+    }
+    co_return latest_offset;
+}
+} // namespace
+
+processor::processor(
+  model::transform_id id,
+  model::ntp ntp,
+  model::transform_metadata meta,
+  std::unique_ptr<wasm::engine> engine,
+  error_callback cb,
+  std::unique_ptr<source> source,
+  std::vector<std::unique_ptr<sink>> sinks,
+  probe* p)
+  : _id(id)
+  , _ntp(std::move(ntp))
+  , _meta(std::move(meta))
+  , _engine(std::move(engine))
+  , _source(std::move(source))
+  , _sinks(std::move(sinks))
+  , _error_callback(std::move(cb))
+  , _probe(p)
+  , _task(ss::now())
+  , _logger(tlog, ss::format("{}/{}", meta.name(), _ntp.tp.partition)) {}
+
+ss::future<> processor::start() {
+    try {
+        co_await _engine->start();
+        co_await _engine->initialize();
+    } catch (const std::exception& ex) {
+        vlog(_logger.warn, "error starting processor engine: {}", ex);
+        _error_callback(_id, _ntp, _meta);
+    }
+    _task = run_transform_loop();
+}
+
+ss::future<> processor::stop() {
+    _as.request_abort();
+    co_await std::exchange(_task, ss::now());
+    co_await _engine->stop();
+}
+
+ss::future<> processor::run_transform_loop() {
+    try {
+        co_await do_run_transform_loop();
+    } catch (const ss::abort_requested_exception&) {
+        // do nothing, we wanted to stop.
+    } catch (...) {
+        vlog(
+          _logger.warn,
+          "error running processor: {}",
+          std::current_exception());
+        _error_callback(_id, _ntp, _meta);
+    }
+}
+ss::future<> processor::do_run_transform_loop() {
+    auto offset = co_await _source->load_latest_offset();
+    vlog(_logger.trace, "starting at offset {}", offset);
+    // TODO(rockwood): Optimize this loop, we should not wait for the transform
+    // or for writing before attempting to read more data.
+    while (!_as.abort_requested()) {
+        auto reader = co_await _source->read_batch(offset, &_as);
+        auto latest_offset = co_await consume_batches(
+          std::move(reader).release(),
+          [this](model::record_batch_reader::data_t batches) {
+              return transform_batches(std::move(batches));
+          },
+          &_as);
```

Comment:
i would have expected here to see something more like `std::move(reader).consume(consumer)` where the consumer called transform batches and finally returned the latest offset. any reason for this alternative pattern?

Reply:
Yes this is adopted from something more complex - I'll simplify - sorry for the noise.